akvo/akvo-lumen
# (WIP) Transform caddisfly results to dataset columns

Changing approach to transformations
**Relates to [Better support Caddisfly data #1571]**
This PR comes from https://github.com/akvo/akvo-lumen/pull/1588

## Steps for UI Testing 
- change `backend/dev/resources/dev.edn` :keycloak-url to point `test` environment (ask @tangrammer if you don't know how :) )
- click "+ Dataset" > select "Akvo Flow" > insert "uat1" > select survey "Caddisfly Test" > select sub-survey "Caddisfly Tests" > select "New form"
- that's all! (you can also verify the update Dataset functionality)

alternative cadddisfly survey pointed by Molo
https://eastafrica.akvoflow.org/admin/ ->Tanga Province - Water Quality Pilot->Water Quality Test Pilot->Fomu ya Upimaji wa ubora wa maji, there's data for E. coli





- [ ] **Update release notes if necessary**

So, instead of converting automatically all caddisfly questions to columns, we try here to add a new `transformation` type. Then final users, once that they've imported the dataset, they can find a new UI option to extract caddisfly values (so far the interface option is added to column dataset header)

- [ ] do we need to support more than strings types in this PR?
- [X] add core/extract-caddisfly column transformation operation [[81647dd](https://github.com/akvo/akvo-lumen/pull/1589/commits/81647dd84b5b5cb59a81e94f913a5ad3931f3b9e)]
- [X] add caddisfly image support `:hasImage`... use "Soil Available Nitrogen" for testing functionality
[3c0ef64
](https://github.com/akvo/akvo-lumen/pull/1589/commits/3c0ef647a326aeddac82aac50fd20bdfe072d04f)
```clojure

;; example schema
{:brand "Quantofix 91320",
  :hasImage true,
  :name
  "Water - Test Strips - Quantofix - Phosphate (0 -100 mg/l PO4)",
  :results [{:id 1, :name "Phosphate", :unit "mg/l"}],
  :uuid "39e7ea7e-9884-407d-a99a-ac3b4d9c9599"}

;; example dataset cell value 

{:image "afbb77ec-b4d0-411a-b833-52f0984207c5.png",
 :name "Water - Fluoride",
 :result [{:id 1, :name "Fluoride", :unit "ppm", :value "> 1.80"}],
 :uuid "f0f3c1dd-89af-49f1-83e7-bcc31c3006cf"}
```  
- [ ] perhaps UI should offer the caddisfly tests to be extracted, for doing so, we could take 2 approachs: 
 + sending to FE the column info related (inside columns data response, if column is caddisfly data type should be extended)
 + allowing the FE access to caddisfly [online schema](https://github.com/akvo/akvo-flow/blob/3fbfed519fd751ea2ab57dd0c838df0cbbde6983/GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java#L38) for taking its own decisions 

- [X] [ [fd5295e](https://github.com/akvo/akvo-lumen/pull/1589/commits/fd5295e713192bc1c1c1be1d9425666b7bbd81b5) ] Add changes for updating older datasets that contains caddisfly questions. 

## functionality available only for new imported or new updates from flow
**We need to enforce the user to update the dataset to have this option thus older lumen dataset doens't contain any caddisfly column data persisted (in lumen)**. ~~Maybe we can do this job in the background so all older columns datasets are automatically updated with caddisfly data (🤔 we need to discuss this option with @janagombitova  before releasing this feature)~~  After checking with @janagombitova we don't want to do any backend (automatically in background) work that implies flow communications :)
